### PR TITLE
Add `internalFormat` to Texture

### PR DESCRIPTION
**Pull request by DavidPeicho**
* Adds `internalFormat` attribute to `Texture` Object;
* Adds `internalFormat` copy to `Texture.copy()`;
* Adds internal formats to constants.js;
* Adds internal formats conversion to `WebGLUtils`;
* Adds singed / unsigned `SAMPLER_3D`, and `SAMPLER_2D_ARRAY` to `WebGLUniforms`

Sorry if sometimes the name are not consistant between each internal format. Just tell me what you like the most between `R16FFormat` and `R16FloatFormat` for instance, it is quite quick to change.

@takahirox sorry for the time I took for that. My own fork already had this feature, and my work was a priority.

### CODE REVIEW

File: src/renderers/webgl/WebGLTextures.js [hunk 1]
```diff
@@ -100,10 +100,13 @@ function WebGLTextures( _gl, extensions, state, properties, capabilities, utils,
 
 	}
 
-	function getInternalFormat( glFormat, glType ) {
+	function deduceInternalFormat( textureInternalFormat, glFormat, glType ) {
```
Comment (Mugen87) [inline src/renderers/webgl/WebGLTextures.js]: I think I would keep the name `getInternalFormat()`. The term `deduce` feels a bit uncommon for the library.

**Review by Mugen87 — CHANGES_REQUESTED** [review]

File: src/renderers/webgl/WebGLTextures.js (hunk 1 above)
Comment (DavidPeicho) [reply src/renderers/webgl/WebGLTextures.js]: Yes but `getInternalFormat` taking itself as an argument the internalFormat was a bit strange. Up to you!

File: src/textures/Texture.js [hunk 2]
```diff
@@ -299,6 +300,12 @@ Texture.prototype = Object.assign( Object.create( EventDispatcher.prototype ), {
 
 		return uv;
 
+	},
+
+	setInternalFormat: function(internalFormat) {
```
Comment (WestLangley) [inline src/textures/Texture.js]: I do not think we need this method. I disagree with @mrdoob. The exception is if it is needed for chaining in a coding pattern we recommend.

I think this method should be removed -- and discussed in a separate PR if you want.

Such methods should return `this` for chaining support, anyway.

Comment (DavidPeicho) [reply src/textures/Texture.js]: Yes forgot the `this`.
As it is an "advanced" usage, just changing the attribute manually is also fine I would say.

**Comment by WestLangley** [comment]
What is the reason for this method?
```javascript
setInternalFormat: function ( internalFormat ) {
    this.internalFormat = internalFormat;
    return this;
}
```
Are you going to add methods for more properties?
```javascript
setGenerateMipmaps()
setPremultiplyAlpha()
setFlipY()
setUnpackAlignment()
// etc...
```
I surely hope not.

**Comment by DavidPeicho** [comment]
This was a request made by @mrdoob, as you commented:
> I do not think we need this method. I disagree with @mrdoob. 

Whenever everyone will agree on this we can make the change, meaning deleting it is needed.

**Comment by DavidPeicho** [comment]
It has been a long time. I am going to fix the linting.
Regarding the setter, I am going to remove it and use the public attribute. If this causes trouble we can still talk about that in another PR as @WestLangley said.

**Comment by bilgorajskim** [comment]
Many formats are still missing, for example R16UI. Here is a table showing all possible formats:

https://webgl2fundamentals.org/webgl/lessons/webgl-data-textures.html

It would be good to have them all.

**Comment by DavidPeicho** [comment]
Good point, the spec is a bit fuzzy on all the available formats / internal formats.
If you want to improve this PR that's even better, you just need to add them in `src/constants.js` and `src/renderers/webgl/WebGLUtils.js`.

**Comment by DavidPeicho** [comment]
Thanks @bilgorajskim for the thorough list of format.

File: src/textures/Texture.js [hunk 3]
```diff
@@ -45,6 +45,7 @@ function Texture( image, mapping, wrapS, wrapT, magFilter, minFilter, format, ty
 	this.anisotropy = anisotropy !== undefined ? anisotropy : 1;
 
 	this.format = format !== undefined ? format : RGBAFormat;
+	this.internalFormat = undefined;
```
Comment (WestLangley) [inline src/textures/Texture.js]: I think the three.js convention would be to default this to `null`.

File: src/renderers/webgl/WebGLUtils.js [hunk 4]
```diff
@@ -51,7 +51,72 @@ function WebGLUtils( gl, extensions, capabilities ) {
...
+		if ( p === RGBA16IFormat ) return gl.RGBA16I;
+		if ( p === RGBA16UIFormat ) return gl.RGBA16UI;
+		if ( p === RGBA16FFormat ) return gl.RGBA16F;
+		if ( p === RGBA32IFormat ) return gl.RGBA32I;
+		if ( p === RGBA32UIFormat ) return gl.RGBA32UI;
+		if ( p === RGBA32FFormat ) return gl.RGBA32F;
+		if ( p === RGB5_A1Format ) return gl.RGB5_A1;
+		if ( p === RGB10_A2Format ) return gl.RGB10_A2;
+		if ( p === RGB10_A2UIFormat ) return gl.RGB10_A2UI;
+		if ( p === SRGB8_ALPHA8Format ) return gl.SRGB8_ALPHA8;
```
Comment (WestLangley) [inline src/renderers/webgl/WebGLUtils.js]: For now, these if-statements should be added at the end; certainly not prior to `AddEquation`.

Perhaps we should implement this method more efficiently in a later PR.

Comment (DavidPeicho) [reply src/renderers/webgl/WebGLUtils.js]: I moved the internal format down. but I let the initial if-statements format up, where they were initially. I am not sure what makes the most sense for you

File: src/renderers/webgl/WebGLTextures.js [hunk 5]
```diff
@@ -118,10 +118,13 @@ function WebGLTextures( _gl, extensions, state, properties, capabilities, utils,
 
 	}
 
-	function getInternalFormat( glFormat, glType ) {
+	function getInternalFormat( internalFormat, glFormat, glType ) {
 
 		if ( ! capabilities.isWebGL2 ) return glFormat;
 
+		// The user provided an internal format, no need to apply deduction.
+		if ( internalFormat ) return utils.convert( internalFormat );
```
Comment (WestLangley) [inline src/renderers/webgl/WebGLTextures.js]: if ( internalFormat !== null )

Zero, if we used it, would be a problem.

**Comment by bilgorajskim** [comment]
I've opened another PR in your fork that fixes a few things that were still missing. @DavidPeicho

**Comment by DavidPeicho** [comment]
Thanks again @bilgorajskim for all the samplers and GL constants addition.

I rebased the branch against the dev branch. What do you think? @WestLangley @Mugen87

File: src/renderers/webgl/WebGLUniforms.js [hunk 6]
```diff
@@ -597,6 +613,50 @@ function setValueT1Array( gl, v, textures ) {
...
 }
 
+function setValueT2DArray1Array( gl, v, textures ) {
+
+	var cache = this.cache;
+	var n = v.length;
+
+	var units = allocTexUnits( textures, n );
+
+	if ( arraysEqual( cache, units ) === false ) {
```
Comment (Mugen87) [inline src/renderers/webgl/WebGLUniforms.js]: For performance reasons, array uniforms are not cached anymore, see #16366

File: src/renderers/webgl/WebGLUniforms.js [hunk 7]
```diff
@@ -597,6 +613,50 @@ function setValueT1Array( gl, v, textures ) {
 
 }
 
+function setValueT2DArray1Array( gl, v, textures ) {
```
Comment (Mugen87) [inline src/renderers/webgl/WebGLUniforms.js]: Why have you included `setValueT2DArray1Array()` and `setValueT3D1Array()` into this PR? It seems to be something different than `Add `internalFormat` to Texture`.

File: src/renderers/webgl/WebGLUniforms.js (hunk 6 above)
Comment (DavidPeicho) [reply src/renderers/webgl/WebGLUniforms.js]: Fixed by removing the uniform setter that shouldn't have been part of this PR anyway.

File: src/renderers/webgl/WebGLUniforms.js (hunk 7 above)
Comment (DavidPeicho) [reply src/renderers/webgl/WebGLUniforms.js]: You are right. Sometimes you know you think about something and you get too crazy about it. Removed and I will do another PR for all of that.

**Comment by DavidPeicho** [comment]
@Mugen87 updated. Mainly removed stuff that shouldn't have been part of this PR, sorry about that.

**Comment by DavidPeicho** [comment]
Just added:
* `internalFormat` to the `Texture` API documentation;
* `PixelFormatInternal` enum to `constants.d.ts` (missing the original formats too, because most of them can be used as internalFormat I think);
* Updates `Textures.d.ts` with the new `PixelFormatInternal` enum for the `internalFormat` class attribute.

Do you think the name I gave to every constants is fine for you? like `RedIntegerFormat`, `RG16IFormat`, `R8_SNORMFormat`, etc... (I basically followed the spec on that).

If the constants names are fine for you, I will update the `Constants` page in the doc.

**Comment by Mugen87** [comment]
> Do you think the name I gave to every constants is fine for you? 

I would say yes 😊 

> The problem right now is that internal formats are typed with the enum PixelFormatInternal, but the format PixelFormat are also available as internal formats. So right now I simply used a TS Union.

I have not much experience with TS but I think we can start with your union approach. So `internalFormat: PixelFormat | PixelFormatInternal;` seems okay to me.

**Comment by DavidPeicho** [comment]
@Mugen87 

* Branch has been rebased on the latest develop;
* All internal formats have been added to the `constants` page in the doc.

I gave a quick description of all the internal format constants. That's like ~50 constants to document, but sometimes their name are a bit shady so that can be good IMO .

**Comment by WestLangley** [comment]
Thank you @DavidPeicho for your persistence. :-) I hope a decision on this can be made soon.

**Review by Mugen87 — APPROVED** [review]

**Comment by Mugen87** [comment]
I would say we merge this at the beginning of the next `dev` cycle `R109`.

**Comment by DavidPeicho** [comment]
I know it's sensitive to merge it because it touches the code of `WebGLTexture` as-well as `WebGLUniforms`, but is there anything I can do to help this one get merged?

**Comment by WestLangley** [comment]
@Mugen87 said

>I vote to merge this at the beginning of the next dev cycle R109.

@mrdoob Can you please have a look?

**Comment by DavidPeicho** [comment]
@mrdoob Do you think this could land in **r111**?

**Comment by mrdoob** [comment]
Oh dear... More than one year old... 😟Sorry for the huge delay! 

Unfortunately this needs proper review and `r111` is due tomorrow.

I'll make sure to have a look at the beginning of `r112` though!

**Comment by WestLangley** [comment]
@Mugen87 Do you think we could have a "three.js extension" that could support this? In other words, support internal format, but only for so-called common formats. The extension could add support for the rest.

I'm just trying to think of a way to make progress here...

**Comment by Mugen87** [comment]
I'm not sure I understand your approach, sorry 😅. What do you mean by "three.js extension"?

I think it's best if we just merge the PR at the beginning of the next dev cycle. `R112` already has many breaking changes.

**Comment by WestLangley** [comment]
>What do you mean by "three.js extension"?

Not sure. I was just trying to promote thinking outside the box.

Maybe you can think of something that is acceptable to @mrdoob.

**Comment by DavidPeicho** [comment]
I haven't been tagged 😇, but same as @Mugen87 I am not sure why this feature would be behind another abstraction layer / feature selector / flag. Three.js already provides `format`, and `type`, so I would say `internalFormat` makes sense to be implemented alongside all of that.

As a sidenote, this *shouldn't* have any breaking changes, because if the `internalFormat` isn't provided, it will fallback to the [original internal format deduction](https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLTextures.js#L140) that was there.

**Comment by WestLangley** [comment]
>so I would say internalFormat makes sense to be implemented alongside all of that.

@WestLangley said

>In other words, support internal format, but only for so-called common formats.

If @mrdoob felt comfortable merging this then he would have done it. I am eliciting suggestions for alternative implementations that may be more acceptable to @mrdoob.

**Comment by mrdoob** [comment]
Sorry for the delay. /me reads

**Comment by mrdoob** [comment]
@DavidPeicho do you have some spare time for resolving the conflicts?

**Comment by DavidPeicho** [comment]
@mrdoob, I just rebased and fix conflicts, should be good.

@WestLangley

What benefits would it bring to split the internal formats modification into _common formats_ (via `Texture`) and an _extension_? It's already transparent for users that don't know anything about WebGL, and expressive enough for people that know WebGL. 

The goal of this feature is **especially** to support more advanced internal format, either to optimize GPU storage, just to get more performance for read / write, or simply to support medical images almost as-is (impossible as-of today without tonemapping the texture 3d, and you don't want that most of the time). Knowing this, when people will try to change the `internalFormat` property, it may often be to use something exotic.

If there is a good rational about it, then I would be glad to hear it and we can figure something out. Otherwise, it's tedious to change this PR again if it doesn't bring anything.

**Comment by WestLangley** [comment]
@DavidPeicho @Mugen87 If mrdoob accepts this, then that is great.

Explaining myself for the _third_ time, I was only raising the issue as an alternative to try to _help you_ get something acceptable merged in the event this PR was not accepted.

**Comment by DavidPeicho** [comment]
The thing is that @mrdoob hasn't given any opinion yet, or at least not here. So either you are aware of  things I don't regarding this particular PR and its content, or we are simply speculating. If this PR is refused then we can think about how to improve it, and what makes most sense in Three.js.

As a side-note, I am totally up to think about the best possible implementation. I know you guys spend a lot of time reviewing / fixing issues, and the work you give for the open source community is great. I am just wondering why, on a 1 year old PR, the foundation of the implementation starts to be questioned only when close to merging and everything is _"ready"_.

**Comment by WestLangley** [comment]
>I am just wondering why, on a 1 year old PR, the foundation of the implementation starts to be questioned

The foundation of the implementation has not been questioned. I have supported you and this PR from the beginning.

I started discussing alternatives _because_ this had not been merged. I think doing so was reasonable, but I will not discuss it further.

**Comment by mrdoob** [comment]
Seems like it was just a misunderstanding 🤝

**Comment by mrdoob** [comment]
The reason this took so long is because I'm worried the library is getting too big and PRs that add code to core scare me.

File: src/renderers/webgl/WebGLUtils.js [hunk 8]
```diff
@@ -147,6 +147,70 @@ function WebGLUtils( gl, extensions, capabilities ) {
 
 		}
 
+		if ( p === RedIntegerFormat ) return gl.RED_INTEGER;
```
Comment (mrdoob) [inline src/renderers/webgl/WebGLUtils.js]: Can we group these inside a `if ( isWebGL2 ) {`?

**Comment by mrdoob** [comment]
Considering that `internalFormat` is pretty much GL...

How about doing:

```js
texture.internalGLFormat = 'R8'
```

And make the code just do `gl[ texture.internalGLFormat ]`?

**Comment by DavidPeicho** [comment]
I would be really up for that, because it makes debugging so much easier. Though right now `format` and `type` are custom Three.js constants mapped to GL, so this doesn't _"compliant"_ in with what has been made.

I am not too opinionated about that in general. In the same ideas of what you propose, what about
```javascript
texture.gpuFormat
```
So it can match later with a `WebGPU` renderer?

**Comment by mrdoob** [comment]
> Though right now `format` and `type` are custom Three.js constants mapped to GL, so this doesn't _"compliant"_ in with what has been made.

That's okay.

> I am not too opinionated about that in general. In the same ideas of what you propose, what about
> ```javascript
> texture.gpuFormat
> ```
> So it can match later with a `WebGPU` renderer?

Sounds good to me 👍

**Comment by mrdoob** [comment]
@DavidPeicho do you think you can update the PR with that change? 😇

**Comment by DavidPeicho** [comment]
@mrdoob Yes I will change it. Do we still go through the mapping in `WebGLUtils` or do we use it like:
```javascript
// User sets the internal format using the GL name.
texture.gpuFormat= 'R8'
```
```javascript
// Renderer uploads the texture using:
var glInternalFormat = WebGL2RenderingContext[texture.gpuFormat]
```

**Comment by mrdoob** [comment]
Those two lines should be enough I think.

**Comment by GuDuJian-J-Zhang** [comment]
hope to merge soon

**Comment by DavidPeicho** [comment]
* Renamed `internalFormat` to `gpuFormat`. Includes:
  * Changing the attribute name in `Texture.js`
  * Modifying slightly the `getInternalFormat()` function in `WebGLTextures.js` to accommodate this change.
  * Updating docs to reflect the new name

I didn't change the current Three.js constant mapping to WebGL2 constants, but I kept it as-is, i.e:
```javascript
myTexture.gpuFormat = THREE.R11F_G11F_B10FFormat; // THREE.R11F_G11F_B10FFormat = 1067
// 1067 is later mapped to WebGL2RenderingContext.R11F_G11F_B10F
```
For a simple reason, the [WebGL2 Specification](https://www.khronos.org/registry/webgl/specs/latest/2.0/) states that you can still use the `RGB` and `RGBA` formats as internal formats. So it means that a user will not be able to do:
```javascript
myTexture.gpuFormat = THREE.RGBFormat;
```

Except if in `WebGLTextures.js` we hardcode something... which didn't seem great like that.

Otherwise we can also delete all constants for gpu format and simply use string, but [WebGPU](https://gpuweb.github.io/gpuweb/#texture-formats) will anyway supports those kind of formats, so I guess it's actually better for Three.js to define it's own constants, and later perform the mapping for each GPU backend (like it's done right now with every constant).

Tell me what you think @mrdoob. If we keep the constants, maybe it could make sense to rename them a bit, I didn't think too much on that.

**Comment by mrdoob** [comment]
> I didn't change the current Three.js constant mapping to WebGL2 constants, but I kept it as-is, i.e:
> ```javascript
> myTexture.gpuFormat = THREE.R11F_G11F_B10FFormat; // THREE.R11F_G11F_B10FFormat = 1067
> // 1067 is later mapped to WebGL2RenderingContext.R11F_G11F_B10F
> ```
> For a simple reason, the [WebGL2 Specification](https://www.khronos.org/registry/webgl/specs/latest/2.0/) states that you can still use the `RGB` and `RGBA` formats as internal formats. So it means that a user will not be able to do:
> ```javascript
> myTexture.gpuFormat = THREE.RGBFormat;
> ```

Then can't do `myTexture.gpuFormat = 'RGB';`?

What I'm proposing is to not add all these new constants and just pass a string to `gpuFormat`.

**Comment by DavidPeicho** [comment]
@mrdoob done.

* Updated `gpuFormat` to a string attribute, matching one of the internal format defined in the [WebGL2 Specification](https://www.khronos.org/registry/webgl/specs/latest/2.0/).
* Added `console.warn` in `WebGLTextures.js` when the specified `gpuFormat` isn't found
* Removed every internal format constants from `constants.js` and `WebGLUtils.js`
* Updated docs to reflect the change from constants to string.
* Updated `Texture.d.ts` and `constants.d.ts` to reflect the `string` type

File: src/renderers/webgl/WebGLUtils.js [hunk 9]
```diff
@@ -147,6 +147,8 @@ function WebGLUtils( gl, extensions, capabilities ) {
 
 		}
 
+		return 0;
```
Comment (mrdoob) [inline src/renderers/webgl/WebGLUtils.js]: Is this intentional?

File: src/textures/Texture.js [hunk 10]
```diff
@@ -110,6 +111,7 @@ Texture.prototype = Object.assign( Object.create( EventDispatcher.prototype ), {
 		this.anisotropy = source.anisotropy;
 
 		this.format = source.format;
+		this.internalFormat = source.internalFormat;
```
Comment (mrdoob) [inline src/textures/Texture.js]:
```js
this.gpuFormat = source.gpuFormat;
```

File: utils/build/rollup.config.js [hunk 11]
```diff
@@ -99,6 +103,51 @@ function glconstants() {
 		TEXTURE_CUBE_MAP_POSITIVE_X: 34069,
 		MAX_CUBE_MAP_TEXTURE_SIZE: 34076,
 		COMPRESSED_TEXTURE_FORMATS: 34467,
+		RED_INTEGER: 36244,
```
Comment (mrdoob) [inline utils/build/rollup.config.js]: We probably don't need all these anymore?

File: utils/build/rollup.config.js [hunk 12]
```diff
@@ -99,6 +103,51 @@ function glconstants() {
...
+		RGBA8I: 36238,
+		RGBA8UI: 36220,
+		RGBA16I: 36232,
+		RGBA16UI: 36214,
+		RGBA32I: 36226,
+		RGBA32UI: 36208,
+		RGB5_A1: 32855,
+		RGB10_A2: 32857,
+		RGB10_A2UI: 36975,
+		SRGB8_ALPHA8: 35907,
```
Comment (mrdoob) [inline utils/build/rollup.config.js]: We probably don't need all these anymore?

File: src/textures/Texture.js (hunk 10 above)
Comment (WestLangley) [reply src/textures/Texture.js]: How about calling it `.internalFormat` consistently, and crossing the WebGPU bridge when we get to it?

Comment (mrdoob) [reply src/textures/Texture.js]: I thought `.gpuFormat` was a good name.
`.internalFormat` can be confusing. Is it internal in Three.js? Or in WebGL?

Comment (WestLangley) [reply src/textures/Texture.js]: It is the internal format as defined in the webGL spec. Personally, I would adhere to that nomenclature.

File: src/renderers/webgl/WebGLUtils.js (hunk 9 above)
Comment (DavidPeicho) [reply src/renderers/webgl/WebGLUtils.js]: It was there I think when I diverged to create my branch, see [here](https://github.com/DavidPeicho/three.js/blob/3585caf4f5779c5eb433edd2838257cd65113e33/src/renderers/webgl/WebGLUtils.js#L158).

I guess it's not needed anymore and I miss that when rebasing. Thanks.

File: src/textures/Texture.js (hunk 10 above)
Comment (DavidPeicho) [reply src/textures/Texture.js]: I fixed that anyway in the last commit.

`internalFormat` also makes more sense to me, but `gpuFormat` is acceptable I guess. Let me know which one you want.

Comment (mrdoob) [reply src/textures/Texture.js]: Lets go with `internalFormat` then.

**Comment by DavidPeicho** [comment]
@mrdoob 

* Renamed `gpuFormat` back to `internalFormat` in `Texure.js`, `WebGLTextures.js`
* Updated docs with `internalFormat`
* Updated `Texture.d.ts` with `internalFormat` as well

Do you want me to squash the commits?

**Comment by Mugen87** [comment]
> Do you want me to squash the commits?

That would be great!

**Comment by DavidPeicho** [comment]
Squashed the commits and rebased.

Just a note to @bilgorajskim, sorry your commits went away. The decision has been made to use directly the GL string as `interalFormat` anyway. Thanks a lot for the help you provided with all the constants!

**Comment by bilgorajskim** [comment]
No problem :) The way it works now is simpler, and I agree with the decision.

**Comment by DavidPeicho** [comment]
@mrdoob 

I pushed again a few changes to `constants.js` and to `WebGLUtils.js`. I forgot about that, but some GL internal formats **require** new GL formats as well. It's just 5 new constants. Unfortunately if we don't add them it makes some nice internal formats unavailable. 

According to the [spec](https://www.khronos.org/registry/webgl/specs/latest/2.0/),
- format `RED_INTEGER` is required to use internal format `R8UI`
- format `RG` is required to use internal formats `RG*`
- format `RG_INTEGER` is requried to use internal format `RG8UI`
- format `RGB_INTEGER` is requried to use internal format `RGB8UI`
- format `RGBA_INTEGER` is requried to use internal format `RGBA8UI`

**Comment by mrdoob** [comment]
Great! I'll do some additional clean up after merging but this looks great.

**Comment by mrdoob** [comment]
Thanks for being patient! 😇

**Comment by mrdoob** [comment]
@DavidPeicho https://github.com/mrdoob/three.js/commit/a8cc32b19b8b2b633fe19eeb06af06bd56892695

**Comment by WestLangley** [comment]
@DavidPeicho Awesome!

**Comment by mrdoob** [comment]
Moving properties around in the constructor is a pain and the usage is messy.
Better to keep constructor properties to a minimum and modify the desired values after construction.

**Pull request merged**